### PR TITLE
rewrites IAA-code

### PR DESCRIPTION
**Pull request by Amelia0010**
https://github.com/yogstation13/Yogstation-TG/pull/1289

#### Changelog

:cl:  
experimental: Internal Affairs Agents now get more objectives as they complete them.
/:cl:

### CODE REVIEW

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/datum/game_mode/traitor/internal_affairs/make_antag_chance()
+	return FALSE
+
+/datum/antagonist/traitor/internal_affairs/iaa_process()
+	return
+
+/datum/game_mode/traitor/internal_affairs/check_finished(force_ending) //to be called by SSticker
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]: Why don't you put this in ``/datum/antagonist/traitor/internal_affairs/process()``?

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/datum/game_mode/traitor/internal_affairs/make_antag_chance()
+	return FALSE
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]: Why?

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	return
+
+/datum/game_mode/traitor/internal_affairs/check_finished(force_ending) //to be called by SSticker
+	.=..()
+	if(force_ending || !SSticker.setup_done || !gamemode_ready || .) //if the gamemode is already considered finished, if it's not begun or if it's being force ended
+		return .
+
+	var/new_target
+	for(var/datum/mind/traitor in traitors)
+		if(!traitor.current || traitor.current.stat==DEAD)
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]:
```
for(var/M in traitors)
    var/datum/mind/traitor = M
```

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+			continue
+		new_target = TRUE
+		for(var/datum/objective/O in traitor)
+			if(O.check_completion() || istype(O, /datum/objective/escape) || istype(O, /datum/objective/survive))
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]:
```
for(var/T in traitor.objectives)
    var/datum/objective/O = T
```

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+				continue
+			new_target = FALSE
+
+		if(new_target)
+			var/list/other_traitors = traitors
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]: Can't you do ``var/list/other_traitors = traitors - traitor`` here and delete the ``other_traitor == traitor`` check a few lines below this?

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+	var/new_target
+	for(var/datum/mind/traitor in traitors)
+		if(!traitor.current || traitor.current.stat==DEAD)
+			continue
+		new_target = TRUE
+		for(var/datum/objective/O in traitor)
+			if(O.check_completion() || istype(O, /datum/objective/escape) || istype(O, /datum/objective/survive))
+				continue
+			new_target = FALSE
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]: So if they have an uncompleted objective, it adds a new objective?

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+		if(new_target)
+			var/list/other_traitors = traitors
+
+			for(var/datum/mind/other_traitor in other_traitors)
+				if(other_traitor == traitor || !other_traitor.current || other_traitor.current.stat==DEAD)
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]:
```
for(var/M in other_traitors)
    var/datum/mind/other_traitor = M
```

File: yogstation/code/game/gamemodes/traitor/double_agents.dm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+				else
+					var/datum/objective/assassinate/internal/kill_objective = new
+					kill_objective.owner = traitor
+					kill_objective.target = newTarget
+					kill_objective.update_explanation_text()
+					traitor.objectives += kill_objective
+
+				traitor.announce_objectives()
+
+	return .
```
Comment (ThatLing) [inline yogstation/code/game/gamemodes/traitor/double_agents.dm]: You don't have to explicitly ``return .``, the game does it for you.

**Review by ThatLing — DISMISSED** [review]
Thanks for this descriptive PR description.

File: yogstation/code/game/gamemodes/traitor/double_agents.dm (hunk 1 above)
Comment (Amelia0010) [reply yogstation/code/game/gamemodes/traitor/double_agents.dm]: cause i'm an idiot

**Comment by Amelia0010** [comment]
and the PR body isn't descriptive, cause I link to a descriptive PR body :Å)

**Review by ThatLing — DISMISSED** [review]

**Comment by Amelia0010** [comment]
is that approval for testmerge @ThatLing?

**Comment by ThatLing** [comment]
What could go wrong

**Pull request merged**